app/src/main/java/com/limelight/Game.java

```diff
@@ -575,8 +575,13 @@ public boolean onKeyDown(int keyCode, KeyEvent event) {
         }
 
         boolean handled = false;
-        if (event.getDevice() == null ||
-                event.getDevice().getKeyboardType() != InputDevice.KEYBOARD_TYPE_ALPHABETIC) {
+
+        boolean detectedGamepad = event.getDevice() == null ? false :
+                ((event.getDevice().getSources() & InputDevice.SOURCE_JOYSTICK) != 0 ||
+                        (event.getDevice().getSources() & InputDevice.SOURCE_GAMEPAD) != 0);
+        if (detectedGamepad || (event.getDevice() == null ||
+                event.getDevice().getKeyboardType() != InputDevice.KEYBOARD_TYPE_ALPHABETIC
+        )) {
```

Comment:
Hey @cgutman, 
 yea it seems to be that the controller has multiple different types. 

To get to the bottom of this, I added the following code 
````
 @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        Log.e("TRACKING","KeyDown-Event: "+event);
        Log.e("TRACKING","KeyDown-Device: "+event.getDevice());
````

The following is what I get out when press the A button

```
05-15 11:59:45.060 12101-12101/com.limelight E/TRACKING: KeyDown-Event: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_BUTTON_A, scanCode=304, metaState=0, flags=0x8, repeatCount=0, eventTime=33390339, downTime=33390339, deviceId=11, displayId=-1, source=0x501 }
05-15 11:59:45.061 12101-12101/com.limelight E/TRACKING: KeyDown-Device: Input Device 11: PG-9068
                                                           Descriptor: 3c74094402619380fa422819a3978fd80e903ecf
                                                           Generation: 166
                                                           Location: external
                                                           Keyboard Type: alphabetic
                                                           Has Vibrator: false
                                                           Has mic: false
                                                           **Sources: 0x1002513 ( keyboard mouse joystick gamepad )**
                                                             AXIS_X: source=0x2002 min=0.0 max=2559.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x2002 min=0.0 max=1439.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_PRESSURE: source=0x2002 min=0.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_VSCROLL: source=0x2002 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_X: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Z: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_RZ: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_GAS: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_BRAKE: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_X: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
05-15 11:59:45.067 12101-12101/com.limelight E/TRACKING: KeyDown-Event: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_DPAD_CENTER, scanCode=304, metaState=0, flags=0x8, repeatCount=0, eventTime=33390339, downTime=33390339, deviceId=11, displayId=-1, source=0x501 }
05-15 11:59:45.068 12101-12101/com.limelight E/TRACKING: KeyDown-Device: Input Device 11: PG-9068
                                                           Descriptor: 3c74094402619380fa422819a3978fd80e903ecf
                                                           Generation: 166
                                                           Location: external
                                                           Keyboard Type: alphabetic
                                                           Has Vibrator: false
                                                           Has mic: false
                                                           Sources: 0x1002513 ( keyboard mouse joystick gamepad )
                                                             AXIS_X: source=0x2002 min=0.0 max=2559.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x2002 min=0.0 max=1439.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_PRESSURE: source=0x2002 min=0.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_VSCROLL: source=0x2002 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_X: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Z: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_RZ: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_GAS: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_BRAKE: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_X: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0

```
And then when I press the B button

````
05-15 13:36:39.934 15135-15135/com.limelight E/TRACKING: KeyDown-Device: Input Device 11: PG-9068
                                                           Descriptor: 3c74094402619380fa422819a3978fd80e903ecf
                                                           Generation: 46
                                                           Location: external
                                                           Keyboard Type: alphabetic
                                                           Has Vibrator: false
                                                           Sources: 0x1002513 ( keyboard mouse joystick gamepad )
                                                             AXIS_X: source=0x2002 min=0.0 max=2559.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x2002 min=0.0 max=1599.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_PRESSURE: source=0x2002 min=0.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_VSCROLL: source=0x2002 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_X: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Z: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_RZ: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_GAS: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_BRAKE: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_X: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
05-15 13:36:39.936 15135-15135/com.limelight E/TRACKING: KeyDown-Event: KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_BACK, scanCode=305, metaState=0, flags=0x8, repeatCount=0, eventTime=67248038, downTime=67248038, deviceId=11, source=0x501 }
05-15 13:36:39.936 15135-15135/com.limelight E/TRACKING: KeyDown-Device: Input Device 11: PG-9068
                                                           Descriptor: 3c74094402619380fa422819a3978fd80e903ecf
                                                           Generation: 46
                                                           Location: external
                                                           Keyboard Type: alphabetic
                                                           Has Vibrator: false
                                                           Sources: 0x1002513 ( keyboard mouse joystick gamepad )
                                                             AXIS_X: source=0x2002 min=0.0 max=2559.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x2002 min=0.0 max=1599.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_PRESSURE: source=0x2002 min=0.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_VSCROLL: source=0x2002 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_X: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_Z: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_RZ: source=0x1000010 min=-1.0 max=1.0 flat=0.11764707 fuzz=0.0 resolution=0.0
                                                             AXIS_GAS: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_BRAKE: source=0x1000010 min=0.0 max=1.0 flat=0.058823533 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_X: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
                                                             AXIS_HAT_Y: source=0x1000010 min=-1.0 max=1.0 flat=0.0 fuzz=0.0 resolution=0.0
````

Here you can see two things, firstly that two different events are fired when i press each button
A results in KEYCODE_BUTTON_A and KEYCODE_DPAD_CENTER
B results in KEYCODE_BUTTON_B and KEYCODE_BACK
and also that the device identifiers as
Sources: 0x1002513 ( keyboard mouse joystick gamepad )

Now I'm not sure why there are two different events going on here. I'm not an android expert so to I followed the turorial on the android developer portal (https://developer.android.com/training/game-controllers/compatibility.html) to see if this is the normal behaviour, however this wasnt happening in this demo and I only got the one event. So not too sure to the root cause of the two events in moonlight is. 

However once I saw that it was skipping over the handling the code when the input device contained a keyboard, this seemed to be the root cause of the issue. Once I implemented this code changes, it has fixed the issue for me across multiple devices so I'm happy to just use it myself, but I think that this change makes sense. 
Why not attempt to handle the inputs if it device is a gamepad? Is there any reason for this logic?
 
If there is a risk associated with this, I could make this a configuration option? 

Regards
Joey